framework/kms-keyring.md

```diff
@@ -309,7 +312,84 @@ If the response is successfully verified, OnDecrypt MUST do the following with t
 If OnDecrypt fails to successfully decrypt any [encrypted data key](#structures.md#encrypted-data-key),
 then OnDecrypt MUST output the unmodified input [decryption materials](#structures.md#decryption-materials).
 
+## Configuration Intent
+
+### OnEncrypt Goal
+
+When a user configures a KMS keyring with key IDs
+and use that keyring to encrypt a message,
+they are stating their intent that they need each one of those CMKs to be able to
+independently decrypt the resulting encrypted message.
+
+For example, if a user configures a KMS keyring with CMKs A, B, and C
+and uses that keyring to encrypt a message,
+then that user expects three things to be true:
+
+1. CMK A can decrypt the encrypted message.
+1. CMK B can decrypt the encrypted message.
+1. CMK C can decrypt the encrypted message.
+
+If any of these are not true of the resulting encrypted message,
+the keyring has failed to honor the user's intent.
+
+In order to accomplish this,
+the KMS keyring MUST contribute three encrypted data keys to the encryption materials:
+one from CMK A, one from CMK B, and one from CMK C.
+
+### OnDecrypt Goal
+
+When a user configures a KMS keyring for use on decrypt,
+they are stating their intent for which CMKs
+the keyring will *attempt* to use to decrypt encrypted data keys.
+
+For example, if a user configures a KMS keyring with CMK C
+and uses it to decrypt an encrypted message
+that contains encrypted data keys for CMKs A, B, and C,
+then the keyring will attempt to decrypt using CMK C.
+
+However, if the keyring attempts to decrypt using CMK C and cannot,
+this failure still honors the configured intent and MUST NOT halt decryption.
+The configured intent is that they keyring MUST *attempt* with these CMKs,
+not that they MUST *succeed*.
+
+### Why OnEncrypt and OnDecrypt are different
+
+On encrypt, the user describes their intent
+for the requirements to decrypt the resulting message.
+Because of this,
+the keyring MUST create encryption materials that satisfies those requirements.
+
+On decrypt, the user provides resources that *attempt* to do that decryption.
+
+This is an asymmetric relationship with very different implications on failure.
+If the keyring encounters a problem on encrypt,
+it cannot fully honor the decryption requirements and so MUST halt message encryption.
+
+However, on decrypt the keyring is not creating anything.
+It is instead attempting to satisfy the requirements that were set on encryption.
+If the keyring cannot satisfy those requirements it MUST NOT halt message decryption.
+
+No keyring can know if it is the last keyring to attempt decryption.
+If all keyrings are exhausted and none of them were able to decrypt an encrypted data key
+then the cryptographic materials manager that managed those keyrings
+will halt message decryption.
```

Comment:
I wonder if some of this belongs in the keyring interface spec, since it isn't specific to the KMS keyring

Reply:
I think we need to be careful here.
There are keyrings that I can envision that do not satisfy this "Any subcomponent can decrypt".
Most notably an N of M style keyring.

Reply:
I'm inclined to say that yes, some form of this should be in the interface spec. Maybe not these exact words, though. I don't think that this PR is the correct place to introduce that, though. I've created #75 to follow up on this.

In the case of either proposed n-of-m keyring, I think these statements still hold true. It is simply the n-of-m keyring that would be making the halt decision rather than one of the child keyrings.

[1] most likely; pending design work

Reply:
I agree that on encrypt describes an intent that on decrypt will satisfy.

I'm saying that in your example of CMKs A, B, C; if I on Encrypt 2 of 3. Then it is not true that A, B, and C call all decrypt the message.  One of them will fail.

This is _not_ the intent for the AWS KMS Keyring, but other keyrings may have such an intent. Just wanting to be clear on how we describe this relationship.
I'll resolve this here, and I think that it is best moved over to #75.

Reply:
I think that will depend on the design of the n-of-m-decrypt keyring. What you described is not what I have in mind for the way such a thing would work, but I agree that this is not the place to design that.